crates/oq3_source_file/src/source_file.rs

```diff
@@ -167,32 +112,37 @@ impl SourceFile {
 }
 
 pub fn search_paths() -> Option<Vec<PathBuf>> {
-    env::var_os("QASM3_PATH").map(|paths| env::split_paths(&paths).collect::<Vec<_>>())
+    env::var_os("QASM3_PATH").map(|paths| env::split_paths(&paths).collect())
 }
 
 /// Expand path with search paths. Return input if expansion fails.
-pub(crate) fn expand_path(file_path: &PathBuf, search_path_list: Option<&Vec<PathBuf>>) -> PathBuf {
+pub(crate) fn expand_path<T: AsRef<Path>, P: AsRef<Path>>(
+    file_path: T,
+    search_path_list: Option<&[P]>,
+) -> PathBuf {
+    let file_path = PathBuf::from(file_path.as_ref());
     if file_path.is_absolute() {
-        return file_path.clone();
+        return file_path;
     }
-    let mut maybe_full_path = file_path.clone();
-    let env_paths = search_paths();
-    // Prefer paths passed to this function over env variable.
-    let paths = if search_path_list.is_some() {
-        search_path_list
-    } else {
-        env_paths.as_ref()
+    let try_path = |dir: &Path| {
+        let full_path = dir.join(&file_path);
+        full_path.is_file().then_some(full_path)
     };
-    if let Some(paths1) = paths {
-        for path in paths1 {
-            let fqpn = path.join(file_path);
-            if fqpn.is_file() {
-                maybe_full_path = fqpn;
-                break;
+
+    if let Some(paths) = search_path_list {
+        for path in paths {
+            if let Some(full_path) = try_path(path.as_ref()) {
+                return full_path;
+            }
+        }
+    } else if let Some(paths) = search_paths() {
+        for path in paths {
+            if let Some(full_path) = try_path(path.as_ref()) {
+                return full_path;
             }
         }
     }
```

Comment:
This isn't ideal with collecting a generic from the user input but also hard-coding access to the environment, but it's well wrapped, and this way doesn't involve collecting a `Vec` from the environment if nothing's given.